AppDB/backup/cassandra_backup.py

```diff
@@ -0,0 +1,308 @@
+
+  Returns:
+    True on success, False otherwise.
+  """
+  logging.info("Restoring Cassandra snapshots.")
+
+  for directory in CASSANDRA_DATA_SUBDIRS:
+    data_dir = "{0}{1}/{2}/".format(constants.APPSCALE_DATA_DIR, "cassandra",
+      directory)
+    logging.debug("Restoring in dir {0}".format(data_dir))
+    for path, _, filenames in os.walk(data_dir):
+      for filename in filenames:
+        logging.debug("Restoring: {0}".format(filename))
+        if not filename:
+          logging.warn("skipping...")
+          continue
+        full_path = "{0}/{1}".format(path, filename)
+        new_full_path = "{0}/../../{1}".format(path, filename)
+        logging.debug("{0} -> {1}".format(full_path, new_full_path))
+        # Move the files up into the data directory.
+        try:
+          call(['cp', full_path, new_full_path])
+        except CalledProcessError as error:
+          logging.error("Error while moving Cassandra snapshot in place. "
+            "Error: {0}".format(str(error)))
+          return False
+  logging.info("Done restoring Cassandra snapshots.")
+  return True
+
+def remove_local_backup_file(local_file=BACKUP_FILE_LOCATION):
+  """ Removes the local backup file.
+
+  Args:
+    local_file: A str, the path to the backup file to delete.
+  """
+  try:
+    call(['rm', '-rf', local_file])
+  except CalledProcessError as error:
+    logging.error("Error while removing local backup file '{0}'. Error: {1}".\
+      format(local_file, str(error)))
+
+def restore_data(storage, path=''):
+  """ Restores the Cassandra snapshot. 
+
+  Args:
+    storage: A str, one of the StorageTypes class members.
+    path: A str, the name of the backup file to restore from.
+  """
+  if storage == StorageTypes.GCS:
+    # Download backup file and store locally with a fixed name.
+    if not gcs_helper.download_from_bucket(path, BACKUP_FILE_LOCATION):
+      logging.error("Download from GCS failed. Aborting recovery...")
+      return False
+
+  # TODO Make sure there's a snapshot to rollback to if restore fails.
+  # Not pressing for fresh deployments.
+  # create_snapshot('rollback-snapshot')
+
+  if not shut_down_cassandra.run():
+    logging.error("Unable to shut down Cassandra. Aborting restore...")
+    clear_old_snapshots()
+    return False
+
+  remove_old_data()
+  try:
+    untar_backup_files()
+  except backup_exceptions.BRException as br_exception:
+    logging.exception("Exception while restoring db snapshots. Need to "
+      "rollback... Exception: {0}".format(str(br_exception)))
+  restore_snapshots()
+
+  # Start Cassandra and repair.
+  start_cassandra.run()
+  refresh_data()
+
+  # Local cleanup.
+  clear_old_snapshots()
+  if storage == StorageTypes.GCS:
+    remove_local_backup_file()
+
+  logging.info("Done with restore.")
+  return True
+
+if "__main__" == __name__:
+  logging.getLogger().setLevel(logging.DEBUG)
+
+  backup_data(storage='', path='')
+  restore_data(storage='', path='')
```

Comment:
What is this doing? If the file is run directly, it performs a backs up and then immediately restores?